File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.py
# 방문할 수 있는 정점이 여러 개인 경우에는 정점 번호가 작은 것을 먼저 방문
# 더 이상 방문할 수 있는 점이 없는 경우 종료한다. 정점 번호는 1번부터 N번까지이다.


# TODO: 그래프를 DFS로 탐색한 결과와 BFS로 탐색한 결과를 출력하는 프로그램을 작성

# n, m, v 입력
# m번 입력

import sys
from collections import deque
input = sys.stdin.readline

n, m, v = map(int, input().split())

graph = [[] for _ in range(n+1)]

# 양방향 간선
for _ in range(m):
    a, b = map(int, input().split())
    graph[a].append(b)
    graph[b].append(a)

for i in range(1, n+1):
    graph[i].sort()


dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]


def dfs(x, visited):
    visited[x] = 1
    print(x, end=' ')
    for i in graph[x]:
        if not visited[i]:
            dfs(i, visited)


def bfs(x):
    queue = deque([x])
    visited = [0 for _ in range(n+1)]
    visited[x] = 1

    while queue:
        x = queue.popleft()

        print(x, end=' ')

        for i in graph[x]:
            if not visited[i]:
                visited[i] = 1
                queue.append(i)


visited = [0 for _ in range(n+1)]

dfs(v, visited)
print()
bfs(v)

File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.js
function solution(n, v, edges) {
  visited = Array(n + 1).fill(false);
  result = [];
  function dfs(x) {
    visited[x] = true;
    result.push(x);

    for (let data of edges[x]) {
      if (!visited[data]) {
        dfs(data);
      }
    }
  }

  dfs(v);
  console.log(result.join(" "));

  function bfs(x) {
    visited = Array(n + 1).fill(false);
    queue = [x];
    answer = [x];

    visited[x] = 1;
    startIdx = 0;

    while (startIdx < queue.length) {
      let x = queue[startIdx];
      startIdx += 1;

      for (let data of edges[x]) {
        if (!visited[data]) {
          queue.push(data);
          answer.push(data);
          visited[data] = true;
        }
      }
    }

    console.log(answer.join(" "));
  }

  bfs(v);
}

/* readline Module */
const readline = require("readline");
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let input;

// n, m, v 입력
// 간선 입력
let n, m, v;
let edges;
let count = 0;

rl.on("line", function (line) {
  if (count === 0) {
    input = line;
    [n, m, v] = input.split(" ").map((el) => Number(el));

    edges = Array(n + 1)
      .fill(new Array())
      .map(() => []);
  } else {
    input = line;
    [a, b] = input.split(" ").map((el) => Number(el));
    edges[a].push(b);
    edges[b].push(a);
  }

  if (m && count == m) {
    rl.close(); // 입력 종료
  }

  count += 1;
}).on("close", function () {
  for (let i = 1; i < edges.length; i++) {
    edges[i].sort((a, b) => a - b);
  }

  solution(n, v, edges); // 문제 풀이 함수 호출
  process.exit(); // 프로세스 종료
});
